data/blog/how-to-refactor-and-not-break-things.md

```diff
@@ -0,0 +1,101 @@
+
+- Be as backward compatible as possible. If possible, our users should not need to change their custom Sentry code.
+- Because we changed how we handle data, we wanted to make sure that we do not use more resources (CPU and memory) compared to before the refactoring.
+- Do not break things. Behavior must stay the same.
+- Do it in baby steps. Have only PRs of manageable size. (No one can review a PR with 120 files changed.)
+- Because this is a massive change, we decided to make it a major version update. Sentry SDK 2.0!
+
+## Phase I: Preparation
+
+There is a quote by [Kent Beck](https://twitter.com/kentbeck/status/250733358307500032):
+
+> Make the change easy, then make the easy change. (Warning, the first part might be hard)
+
+In this first phase, we moved existing functionality away from the Hub into the Scope. We did several PRs that each moved one or two functions from the Hub into the Scope and changed the function in the Hub to call its counterpart in the Scope.
+
+This way we had a couple of smaller PRs that each dealt with one topic and were easy to review.
+
+After we moved all the functionality from the Hub into the Scope our extensive test suite was still all green, giving us confidence that we did not change any behavior. Because neither the top-level Sentry API nor the Hub-based API was changed this could be released in a minor version update.
+
+This concluded the preparation of our canvas. Time to make the change.
+
+## Phase II: New Scopes
+
+Phase I left us with a hollow shell of a Hub and all functionality in the Scope. We now refactored the Scope. This was the biggest part. We changed how the Scope was stored in memory (it's not on the Hub anymore but saved as a Python Context Variable). We also introduced three different flavors of the Scope for better encapsulation of data. If you want to dig into details read our [develop docs on the topic](https://develop.sentry.dev/sdk/hub_and_scope_refactoring/).
+
+After we updated the Scope the old Hub-based API could still be used, but under the hood, the new Scopes-based API was called. We did not release this phase because we wanted to first do Phase III.
+
+Again, our test suite gave us confidence that the SDK still behaved the same as before.
+
+## Phase III: Use New Scopes Everywhere
+
+In the Python SDK, we have over 40 integrations for various web frameworks, databases, and other libraries. All of them still used the Hub-based API.
+
+We updated each integration from the old Hub-based API to the new Scope-based API in a separate PR. This was pretty straightforward and done in a couple of days.
+
+Doing this gave us insights into the look and feel of the new API. Because we now used the API in the same way as our power users would use it in the future. We found some things that we did not like and made some minor changes to the new API to make it more convenient to use. Having one PR for each integration created a lot of small PRs that were easy to review.
```

Comment:
> We found some things that we did not like and made some minor changes to the new API to make it more convenient to use

Would love to see some code block here. What was it before and why didn't u like it?
What was it after?

Reply:
Good idea! Added a code snippet and links to prs!